src/auth/webhook-secret.guard.ts

```diff
@@ -0,0 +1,32 @@
+import {
+  CanActivate,
+  ExecutionContext,
+  Injectable,
+  UnauthorizedException,
+} from '@nestjs/common';
+import * as crypto from 'crypto';
+import { Request } from 'express';
+import { Env } from 'src/env';
+
+@Injectable()
+export class WebhookSecretGuard implements CanActivate {
+  canActivate(context: ExecutionContext): boolean {
+    const request: Request = context.switchToHttp().getRequest();
+    const signature = request.headers['x-hub-signature-256'] as string;
+    const secret = Env.SECRET.WEBHOOK;
+    const payload = JSON.stringify(request.body);
+
+    if (!signature) {
+      throw new UnauthorizedException('Missing signature');
+    }
+
+    const hmac = crypto.createHmac('sha256', secret);
+    const digest = `sha256=${hmac.update(payload).digest('hex')}`;
+
+    if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
```

Comment:
PullProofAI

The current implementation of comparing signatures using `crypto.timingSafeEqual` can potentially leak timing information, making it susceptible to timing attacks. It's recommended to use a constant-time comparison method to mitigate this vulnerability.

```diff
- if (!crypto.timingSafeEqual(Buffer.from(digest), Buffer.from(signature))) {
+ if (!constantTimeCompare(digest, signature)) {
```

You should implement a constant-time comparison function like the one below:

```typescript
function constantTimeCompare(a: string, b: string): boolean {
    if (a.length !== b.length) {
        return false;
    }

    let result = 0;
    for (let i = 0; i < a.length; i++) {
        result |= a.charCodeAt(i) ^ b.charCodeAt(i);
    }

    return result === 0;
}
```
This function ensures that the comparison takes the same amount of time regardless of the input values, reducing the risk of timing attacks.


<!-- This is an auto-generated comment by PullProofAI -->